runtime/lua/vim/fswatch/fswatch.lua

```diff
@@ -0,0 +1,140 @@
+--[[
+  Create a file watcher, each watcher is identified by the name of the file that it
+  watches. We can use a lua table to store all watchers indexed by their filenames
+  so that we can close the required watcher during the callback to on_change to
+  debounce the watcher.
+--]]
+
+local uv = vim.loop
+
+local Watcher = {
+  fname = '',
+  ffname = '',
+  handle = nil,
+  paused = false,
+  pending_notifs = false,
+}
+local WatcherList = {}
+
+-- idle handle to check for any pending notifications in a watcher.
+-- Only displays the notifications if neovim is in focus, and the buffer
+-- is the current buffer.
+
+-- Callback for the idle handle
+function check_notifications()
+  for f, watcher in pairs(WatcherList) do
+    if watcher.pending_notifs and watcher.paused == false then
+      if uv.fs_stat(watcher.ffname) ~= nil then
+        vim.api.nvim_command('call fswatch#PromptReload()')
+      else
+        print("ERR: File "..watcher.fname.." removed")
+      end
+      watcher.pending_notifs = false
+    end
+  end
+end
+
+local check_handle = uv.new_check()
+check_handle:start(vim.schedule_wrap(check_notifications))
+
+function Watcher:new(fname)
+  assert(fname ~= '', 'Watcher.new: Error: fname is an empty string')
+  -- get full path name for the file
+  local ffname = vim.api.nvim_call_function('fnamemodify', {fname, ':p'})
+  w = {fname = fname, ffname = ffname, handle = nil}
+  setmetatable(w, self)
+  self.__index = self
+  return w
+end
+
+function Watcher:start()
+  assert(self.fname ~= '', 'Watcher.start: Error: no file to watch')
+  assert(self.ffname ~= '', 'Watcher.start: Error: full path for file not available')
+  -- get a new handle
+  self.handle = uv.new_fs_event()
+  self.handle:start(self.ffname, {}, self.on_change)
+end
+
+function Watcher:stop()
+  assert(self.fname ~= '', 'Watcher.stop: Error: no file being watched')
+  assert(self.handle ~= nil, 'Watcher.stop: Error: no handle watching the file')
+  self.handle:stop()
+  -- close the handle altogether, for windows.
+  if self.handle:is_closing() then
+    return
+  end
+  self.handle:close()
+end
+
+function Watcher.on_change(err, fname, event)
+  WatcherList[fname].pending_notifs = true
+
+  WatcherList[fname]:stop()
+  WatcherList[fname]:start()
```

Comment:
Why do we have to stop and start the watchers? (it is not a criticism, I just don't remember why)

Reply:
When neovim writes files it does it by moving the original file to somewhere else and moving a new temporary file in its place. Thus the old watcher becomes invalid. We do this so that we can start the watcher again on the same file after it has been replaced.

Reply:
> When neovim writes files it does it by moving the original file to somewhere else and moving a new temporary file in its place.

That depends on how [`'backupcopy'`](https://neovim.io/doc/user/options.html#'backupcopy') is configured.

Reply:
Umm, doesn't it still rename the original file so as to make the write atomic? https://github.com/neovim/neovim/blob/master/src/nvim/fileio.c#L2912

Reply:
Yes, if `backupcopy` is `yes` it simply copies the file instead of renaming it.